ARCHITECTURE.md

```diff
@@ -0,0 +1,140 @@
+job to illustrate the workflow. See the published artifacts for this job [here](https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/openshift_machine-config-operator/112/pull-ci-openshift-machine-config-operator-master-images/30?log#log).
+
+## Input Resolution
+
+In order to ensure that work is not done twice, `ci-operator` needs to determine
+when work can be re-used. The tool will identify a build of any specific job with
+a hash of:
+
+ - job metadata (refs to test, clone configuration)
+ - configuration (YAML configuration for refs under test)
+ - other inputs (levels of input tagged images)
+
+With such an identifier, the `ci-operator` can determine if two builds are using
+the same configuration on the same inputs and can therefore re-use common work.
+This identifier is used to create the Kubernetes `Namespace` in which the test
+workloads will run and is furthermore available to templatized tests with the
+[`$NAMESPACE`](./TEMPLATES.md#namespace) variable.
+
+Input resolution can be identified in the `ci-operator` output by all of the
+steps that precede the creation of the test `Namespace`:
+
+```
+2018/10/08 05:28:45 Resolved source https://github.com/openshift/machine-config-operator to master@dc1d2c7f, merging: #112 f63d7a2c @flaper87
+2018/10/08 05:28:45 Resolved openshift/release:golang-1.10 to sha256:43ad4740f25e4cbc6de8a964e2ce65ce50fbc24fc4e5e268e0a7370ca3b09bd1
+2018/10/08 05:28:45 Resolved openshift/origin-v4.0:base to sha256:903622079f42a79fd6f3bbb6f2419292883b94c3227638fc93a99f77ebc96b00
+2018/10/08 05:28:45 Using namespace ci-op-31xmgx1s
+```
+
+## Namespace Initialization
+
+The hash created from input resolution is used to create a `Namespace` as an
+isolated workspace for the test; the `Namespace` is subsequently initialized for
+use by the test workloads.
+
+All input images for the tests that are described in the configuration YAML are
+tagged in, as are all images that form the larger release that the test is a part
+of. Images that are used for the build graph, like those identified with the
+[`base_images`](./CONFIGURATION.md#base_images),
+[`base_rpm_images`](./CONFIGURATION.md#base_rpm_images), and
+[`build_root`](./CONFIGURATION.md#build_root) stanzas, have `ImageStreamTag`s
+created for them in the `pipeline` `ImageStream` in the test `Namespace`. Images
+that are part of the release that the test exists within, as specified with the
+[`tag_specification`](./CONFIGURATION.md#tag_specification) stanza, are mirrored
+to `ImageStreamTag`s in the `stable` `ImageStream` within the test `Namespace`.
+
+In order to ensure that resources from tests do not leak to the cluster the tests
+are executed on, both hard and soft TTLs are set on the `Namespace` and the
+[`ci-ns-ttl-controller`](https://github.com/openshift/ci-ns-ttl-controller) is
+used to enforce the TTLs and reap namespaces when TTLs have expired.
```

Comment:
We've been asked a few times about this, so it's worth expanding this section to say that the TTL controller is deployed on `api.ci` and describe what happens (or doesn't) when it's not deployed or if the user doesn't have permission to edit the namespace.